### PR TITLE
validate `devcontainer-feature.json` schema

### PR DESCRIPTION
**Pull request by joshspicer**
closes: https://github.com/devcontainers/action/issues/108
ref: https://github.com/devcontainers/spec/issues/140

Take the published `devcontainer-feature.json` schema (https://github.com/devcontainers/spec/pull/152) and uses it to validate the repo contents before publishing, emitting an error if a Feature does not follow the published schema.

Adds unit tests to validate behavior

Example of a schema failure:

https://github.com/codspace/features-with-joshspicer-custom-action/actions/runs/3742904740/jobs/6354427166
image

### CODE REVIEW

**Review by samruddhikhandale — APPROVED** [review]
Looks great, left minor comments.

💡  Maybe in future we could create a workflow with a daily/weekly cadence to auto-create a PR for `src/schemas/devContainerFeature.schema.json` file.

File: src/schemas/README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Schemas
+
+Schemas can be updated by running `yarn update-schemas` from the root of the project.
```
Comment (chrmarti) [inline src/schemas/README.md:3]: Seems to have changed to `fetch-schemas`.

**Review by chrmarti — APPROVED** [review]
LGTM!

Another option to checking in a copy of the schema would be to download it at runtime when it isn't there. (This is only the case when running the action from the repository, not when running it as published action, as far as I see.)

File: src/main.ts [hunk 2]
```diff
@@ -52,6 +54,12 @@ async function run(): Promise<void> {
     }
 
     if (shouldPublishFeatures) {
+        core.info('Validating Feature metadata...');
+        if (!(await prePublish('feature', featuresBasePath))) {
+            core.setFailed('(!) Failed to validate contents of Features.');
```
[the remark this answers is not in the document]
Comment (joshspicer) [reply src/main.ts:59]: good point!

**Pull request merged**